lib/hypothesis/engine.rb

```diff
@@ -6,41 +6,51 @@
 module Hypothesis
   class Engine
     attr_reader :current_source
+    attr_accessor :is_find
 
     def initialize(max_examples: 200, seed: nil)
       seed = Random.rand(2**64 - 1) if seed.nil?
       @core_engine = HypothesisCoreEngine.new(seed, max_examples)
     end
 
     def run
-      while @core_engine.should_continue
+      loop do
         core_id = @core_engine.new_source
-        @current_source = Source.new(@core_engine, core_id)
+        break if core_id.nil?
+        @current_source = Source.new(@core_engine, core_id, false)
         begin
-          yield(@current_source)
+          result = yield(@current_source)
+          @core_engine.finish_interesting(core_id) if is_find && result
         rescue UnsatisfiedAssumption
           @core_engine.finish_invalid(core_id)
         rescue DataOverflow
           @core_engine.finish_overflow(core_id)
         rescue StandardError
+          raise if is_find
           @core_engine.finish_interesting(core_id)
         else
           @core_engine.finish_valid(core_id)
         end
       end
-      raise Unsatisfiable if @core_engine.was_unsatisfiable
       core_id = @core_engine.failing_example
-      return if core_id.nil?
+      if core_id.nil?
+        raise Unsatisfiable if @core_engine.was_unsatisfiable
+        return
+      end
 
-      @current_source = Source.new(@core_engine, core_id)
+      @current_source = Source.new(@core_engine, core_id, is_find)
       yield @current_source
     end
   end
 
   class Source
-    def initialize(core_engine, core_id)
+    attr_reader :draws
+
+    def initialize(core_engine, core_id, record_draws)
       @core_engine = core_engine
       @core_id = core_id
+      @record_draws = record_draws
+      @draws = []
```

Comment:
Is `@draws` only used for recording draws? If so, maybe we could wrap both of these class variables in an if block?

Reply:
Although after rubocop got at it and made me use `&.` I'm not sure it's entirely an improvement. Can I get automatic formatting of Ruby *without* the bad reckons somehow?